libsolidity/ast/Types.cpp

```diff
@@ -1699,7 +1715,7 @@ TypePointer TupleType::closestTemporaryType(TypePointer const& _targetType) cons
 FunctionType::FunctionType(FunctionDefinition const& _function, bool _isInternal):
 	m_location(_isInternal ? Location::Internal : Location::External),
 	m_isConstant(_function.isDeclaredConst()),
-	m_isPayable(_function.isPayable()),
+	m_isPayable(_isInternal ? false : _function.isPayable()),
```

Comment:
This is the constructor that creates a function type out of a function definition. If the function is `public`, it can be payable, and the resulting internal function type will not be payable.